pkg/provider/provider.go

```diff
@@ -621,3 +636,83 @@ func validateFileName(fileName string) error {
 	}
 	return nil
 }
+
+type node struct {
+	cert     *x509.Certificate
+	parent   *node
+	isParent bool
+}
+
+func fetchCertChains(data []byte) ([]byte, error) {
+	var newCertChain []*x509.Certificate
+	var pemData []byte
+	nodes := make([]*node, 0)
+
+	for {
+		// decode pem to der first
+		block, rest := pem.Decode(data)
+		data = rest
+
+		if block == nil {
+			break
+		}
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return pemData, err
+		}
+		nodes = append(nodes, &node{
+			cert:     cert,
+			parent:   nil,
+			isParent: false,
```

Comment:
I'm curious why you need both 'parent' and 'isParent'.  Isn't that redundant.  Can you just check parent for nil?  Or is this a Go idiom?

Reply:
The idea was to do the leaf cert lookup O(n) using `isParent`. If `isParent` is false, then that's the leaf node and then we just construct the chain until we hit the `nil` node.